### PR TITLE
Plugin to verify the username is equal to the certificate CN

### PR DESCRIPTION
**Pull request by stephdl**
Verify the username of the client configuration is equal the common name of the certificate 

https://github.com/NethServer/dev/issues/6111

### CODE REVIEW

**Review by gsanchietti — COMMENTED** [review]
IMO the script behavior is not really clear.
We need to validate the match between the login name and the certificate name.

Which is the AuthMode for this scenario? I'd like to have a more explicit logic, something like (pseudo-code):
```
if (mode == "the one required validation") {
   if (given_login == certificate_cn || given_login_long == certificate_cn) {
       exit 0
   } else {
       exit 1
   }
}

# no check required
exit 0
```

**Comment by stephdl** [comment]
```
if [[ $AUTHMODE = 'certificate' || $AUTHMODE = 'password' ]];then
    exit 0
fi
```

@gsanchietti , rather to clearly states on what authentication mode to verify what username must valid the common name of the certificate, does it sounds better for the future to `exit 0` when the validation means no sense: eg certificate alone, or password alone 

:-?

**Comment by giacomo** [comment]
@stephdl could you please avoid to use my usertag 😂 thanks

**Comment by stephdl** [comment]
sorry it is the second time :)

**Comment by gsanchietti** [comment]
> does it sounds better for the future to `exit 0` when the validation means no sense: eg certificate alone, or password alone

Yes, `exit 0` also means "nothing to do".

The current script seem strange to me, maybe just chaining the conditions, improving the script name and the comment could be enough.
But when I saw it, I was totally lost (and I already knew what was for!)

**Comment by stephdl** [comment]
ok will do some better condition and try to explain better what it does

**Comment by stephdl** [comment]
new version with  perl @gsanchietti

File: root/usr/libexec/nethserver/connect-scripts/openvpn-connect-check-username-equal-certficateCN [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+# NethServer is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with NethServer.  If not, see COPYING.
+#
+
+# When the remote client initates a connection, he sents a $username ENV variable
```
Comment (gsanchietti) [inline root/usr/libexec/nethserver/connect-scripts/openvpn-connect-check-username-equal-certficateCN]:
```suggestion
# When the remote client initates a connection, it sends a $username ENV variable
```

File: root/usr/libexec/nethserver/connect-scripts/openvpn-connect-check-username-equal-certficateCN [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+my $AuthMode = $db->get_prop('openvpn@host-to-net', 'AuthMode') || 'password';
+my $OtpStatus = $db->get_prop('openvpn@host-to-net', 'OtpStatus') || 'disabled';
+my $DomainName = $db->get_value('DomainName') || '';
+
+# we exclude two authentication modes (certificate without OTP & password) 
+if ($AuthMode eq 'certificate' && $OtpStatus eq 'disabled') {
+    exit 0;
+} elsif ($AuthMode eq 'password') {
+    exit 0;
+}
```
Comment (gsanchietti) [inline root/usr/libexec/nethserver/connect-scripts/openvpn-connect-check-username-equal-certficateCN]:
```suggestion
if ( $AuthMode eq 'password' || ($AuthMode eq 'certificate' && $OtpStatus eq 'disabled') ) {
    exit 0;
}
```

File: root/usr/libexec/nethserver/connect-scripts/openvpn-connect-check-username-equal-certficateCN [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+# we valid if the name sent by the openvpn client configuration is equal to the 
+# common name of the certificate (also for short username and long username), 
+# else we exit with an error.
+
+if ($ENV{'common_name'} eq $ENV{'username'}) {
+    exit 0;
+} elsif ($ENV{'common_name'} eq ($ENV{'username'}.'@'.$DomainName)){
+    exit 0;
+}
```
Comment (gsanchietti) [inline root/usr/libexec/nethserver/connect-scripts/openvpn-connect-check-username-equal-certficateCN]:
```suggestion
if ($ENV{'common_name'} eq $ENV{'username'} || $ENV{'common_name'} eq ($ENV{'username'}.'@'.$DomainName)) {
    exit 0;
}
```

**Review by gsanchietti — CHANGES_REQUESTED** [review]
Improve the code by chaining multiple conditions, in this way we can take advantage from https://perlmaven.com/short-circuit

File: root/usr/libexec/nethserver/connect-scripts/openvpn-connect-check-username-equal-certficateCN (hunk 1 above)
Comment (stephdl) [reply root/usr/libexec/nethserver/connect-scripts/openvpn-connect-check-username-equal-certficateCN]: done

File: root/usr/libexec/nethserver/connect-scripts/openvpn-connect-check-username-equal-certficateCN (hunk 2 above)
Comment (stephdl) [reply root/usr/libexec/nethserver/connect-scripts/openvpn-connect-check-username-equal-certficateCN]: done

File: root/usr/libexec/nethserver/connect-scripts/openvpn-connect-check-username-equal-certficateCN (hunk 3 above)
Comment (stephdl) [reply root/usr/libexec/nethserver/connect-scripts/openvpn-connect-check-username-equal-certficateCN]: done

**Review by gsanchietti — APPROVED** [review]
Please merge only when a issue is ready.

**Pull request merged**